### PR TITLE
Forward port of tendermint/tendermint#7605, tendermint/spec#393, and tendermint/tendermint#8142

### LINKED ISSUE #2003: Forward port work present in branch `wb/proposer-based-timestamps`
Around 70% PBTS work was done on branch `wb/proposer-based-timestamps` (as sort of feature branch). However, that branch:
* wasn't merged back on `master` with a merge commit
* was rather "squash-merged" into a big PR: tendermint/tendermint#7605.

This issue forward-ports tendermint/tendermint#7605 on our new feature branch `feature/pbts`.

* [x] #2018
  * Original PR: tendermint/tendermint#7605, this is a big "merge-squash" PR.
    * The original merge-squash PR contains the following smaller PRs (which have thus been forward-ported in this PR)
      * tendermint/tendermint#7170
      * tendermint/tendermint#7305
      * tendermint/tendermint#7300
      * tendermint/tendermint#7290
      * tendermint/tendermint#7307
      * tendermint/tendermint#7341
      * tendermint/tendermint#7373
      * tendermint/tendermint#7354
      * tendermint/tendermint#7376
      * tendermint/tendermint#7382
      * tendermint/tendermint#7418
      * tendermint/tendermint#7391
      * tendermint/tendermint#7413
      * tendermint/tendermint#7480
      * tendermint/tendermint#7415
      * tendermint/tendermint#7554
      * tendermint/tendermint#7563
  * This fw-port PR also includes two other PRs not included in tendermint/tendermint#7605 but needed by it
    * tendermint/spec#393 (protobufs, which were in a separate repo at the time)
    * tendermint/tendermint#8142

---
## DoD

* Forward-port tendermint/tendermint#7605 and all other needed PRs
  * Exclude unnedded refactoring not related to PBTS
* Production code builds
* Unit and e2e tests pass

### PR DESCRIPTION
**Pull request by sergio-mena**
Closes #2003

This is a pure forward-port of tendermint/tendermint#7605, which was a "squashed merge" of several other PRs.

There are several refactorings included in tendermint/tendermint#7605 that have been left out of the cherry-pick (not necessary and their absence makes this PR smaller):
- tendermint/tendermint#6969
- Everything that has to do with context threading, and `RandState` vs. `MakeState` (since the logic this changes is anyway not present on `main` anyway)

Finally, two further commits had to be be forward-ported together with tendermint/tendermint#7605:
- tendermint/spec#393 When PBTS was implemented (end of 2021 - beginning of 2022) the protobufs were held in a different repo (spec). This PR on that repo contains those changes, which are needed by tendermint/tendermint#7605
- tendermint/tendermint#8142 Minor adjustments to Synchrony parameters that was committed separately on `master`

---

#### Original description of tendermint/tendermint#7605:

This pull request merges in the changes for implementing Proposer-based timestamps into `master`. The power was primarily being done in the `wb/proposer-based-timestamps` branch, with changes being merged into that branch during development. This pull request represents an amalgamation of the changes made into that development branch. All of the changes that were placed into that branch have been cleanly rebased on top of the latest `master`. The changes compile and the tests pass insofar as our tests in general pass.

 These changes have been extensively reviewed during development. There is not much new here. In the interest of making effective use of time, I would recommend against trying to perform a complete audit of the changes presented and instead examine for mistakes that may have occurred during the process of rebasing the changes. I gave the complete change set a first pass for any issues, but additional eyes would be very appreciated.

In sum, this change set does the following:
closes #6942
merges in #6849

---

#### PR checklist

- [x] Tests written/updated
- [x] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [x] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: internal/state/state.go [hunk 1]
```diff
@@ -260,29 +259,6 @@ func (state State) MakeBlock(
 	return block
 }
 
-// MedianTime computes a median time for a given Commit (based on Timestamp field of votes messages) and the
-// corresponding validator set. The computed time is always between timestamps of
-// the votes sent by honest processes, i.e., a faulty processes can not arbitrarily increase or decrease the
-// computed value.
-func MedianTime(commit *types.Commit, validators *types.ValidatorSet) time.Time {
```
Comment (glnro) [inline internal/state/state.go:267]: This should be restored per #2063

File: types/params.go [hunk 2]
```diff
@@ -131,6 +142,15 @@ func DefaultABCIParams() ABCIParams {
 	}
 }
 
+func DefaultSynchronyParams() SynchronyParams {
+	// TODO(@wbanfield): Determine experimental values for these defaults
```
Comment (glnro) [inline types/params.go:146]: @sergio-mena can we put together a QA doc and include this

File: types/params_test.go [hunk 3]
```diff
@@ -23,24 +23,176 @@ func TestConsensusParamsValidation(t *testing.T) {
...
 	}{
 		// test block params
-		0: {makeParams(1, 0, 2, 0, valEd25519, 0), true},
-		1: {makeParams(0, 0, 2, 0, valEd25519, 0), false},
-		2: {makeParams(47*1024*1024, 0, 2, 0, valEd25519, 0), true},
-		3: {makeParams(10, 0, 2, 0, valEd25519, 0), true},
-		4: {makeParams(100*1024*1024, 0, 2, 0, valEd25519, 0), true},
-		5: {makeParams(101*1024*1024, 0, 2, 0, valEd25519, 0), false},
-		6: {makeParams(1024*1024*1024, 0, 2, 0, valEd25519, 0), false},
+		{
```
Comment (glnro) [inline types/params_test.go:26]: Might add a testcase name/description

File: types/proposal.go [hunk 4]
```diff
@@ -34,14 +34,14 @@ type Proposal struct {
 
 // NewProposal returns a new Proposal.
 // If there is no POLRound, polRound should be -1.
-func NewProposal(height int64, round int32, polRound int32, blockID BlockID) *Proposal {
+func NewProposal(height int64, round int32, polRound int32, blockID BlockID, ts time.Time) *Proposal {
```
Comment (glnro) [inline types/proposal.go:37]: @sergio-mena we should note that if we backport BFT time, then we'll need to refactor everywhere `NewProposal` is called as well

Comment (sergio-mena) [reply types/proposal.go:37]: Added comment to #2063

File: internal/statesync/messages_test.go [hunk 5]
```diff
@@ -120,3 +121,7 @@ func TestStateSyncVectors(t *testing.T) {
 		require.Equal(t, tc.expBytes, hex.EncodeToString(bz), tc.testName)
 	}
 }
+
+func durationPtr(t time.Duration) *time.Duration {
```
Comment (glnro) [inline internal/statesync/messages_test.go]: I'm not sure why this is necessary

File: spec/abci/abci++_methods.md [hunk 6]
```diff
@@ -749,11 +749,13 @@ Most of the data structures used in ABCI are shared [common data structures](../
...
+    | --------- | ------------------------------------------------------------- | ---------------------------------------------------------------------------- | ------------ | ------------- |
     | block     | [BlockParams](../core/data_structures.md#blockparams)         | Parameters limiting the size of a block and time between consecutive blocks. | 1            | Yes           |
     | evidence  | [EvidenceParams](../core/data_structures.md#evidenceparams)   | Parameters limiting the validity of evidence of byzantine behaviour.         | 2            | Yes           |
     | validator | [ValidatorParams](../core/data_structures.md#validatorparams) | Parameters limiting the types of public keys validators can use.             | 3            | Yes           |
     | version   | [VersionsParams](../core/data_structures.md#versionparams)    | The ABCI application version.                                                | 4            | Yes           |
+    | abci      | [ABCIParams](../core/data_structures.md#abciparams)           | ABCI-related parameters.                                                     | 5            | Yes           |
```
Comment (glnro) [inline spec/abci/abci++_methods.md:757]: Mentioning for tracking but I think this was discussed in our session that this doesn't belong in the pr

File: spec/core/data_structures.md [hunk 7]
```diff
@@ -485,11 +490,24 @@ func SumTruncated(bz []byte) []byte {
 |---------------|-----------------|-----------------------------------------------------------------------|--------------|
 | pub_key_types | repeated string | List of accepted public key types. Uses same naming as `PubKey.Type`. | 1            |
 
+### ABCIParams
+
+| Name                          | Type  | Description                   | Field Number |
+| ----------------------------- | ----- | ----------------------------- | ------------ |
+| vote_extensions_enable_height | int64 | The ABCI application version. | 1            |
```
Comment (glnro) [inline spec/core/data_structures.md:497]:
```suggestion
| vote_extensions_enable_height | int64 | The height where vote extensions will be enabled. | 1            |
```

File: internal/consensus/common_test.go [hunk 8]
```diff
@@ -733,14 +759,6 @@ func ensureVoteMatch(t *testing.T, voteCh <-chan cmtpubsub.Message, height int64
 	}
 }
 
-func ensurePrecommitTimeout(ch <-chan cmtpubsub.Message) {
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/common_test.go:736]: Cherry-picked from tendermint/tendermint#7605, if the UTs build and run well (and the function not being public, that means to me it's not needed)

File: internal/consensus/state.go [hunk 9]
```diff
@@ -1159,29 +1173,29 @@ func (cs *State) enterPropose(height int64, round int32) {
...
-
 	if cs.privValidatorPubKey == nil {
 		// If this node is a validator & proposer in the current round, it will
 		// miss the opportunity to create a block.
 		logger.Error("propose step; empty priv validator public key", "err", ErrPubKeyIsNotSet)
 		return
 	}
 
-	address := cs.privValidatorPubKey.Address()
+	addr := cs.privValidatorPubKey.Address()
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state.go:1187]: I left it as it was in tendermint/tendermint#7605 (don't forget this is a cherry-pick commit).
Besides I find the change good: less characters, and still conveying the same info.

File: internal/consensus/state.go [hunk 10]
```diff
@@ -1346,6 +1369,32 @@ func (cs *State) defaultDoPrevote(height int64, round int32) {
 		return
 	}
 
+	if cs.Proposal == nil {
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state.go:1372]: Any way we can unit-test this?

[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state.go:1372]: Also (now that I remember). I this condition is necessary in order for line 1378 not to panic. In other words, the logic implemented in line 1378 assumes we have a proposal and a proposed block. And that logic seems to me it's core PBTS logic

File: internal/consensus/state.go [hunk 11]
```diff
@@ -1583,19 +1645,19 @@ func (cs *State) enterPrecommit(height int64, round int32) {
 	// the proposed block, update our locked block to this block and issue a
 	// precommit vote for it.
 	if cs.ProposalBlock.HashesTo(blockID.Hash) {
-		logger.Debug("precommit step; +2/3 prevoted proposal block; locking", "hash", blockID.Hash)
+		logger.Debug("precommit step: +2/3 prevoted proposal block; locking", "hash", blockID.Hash)
 
 		// Validate the block.
 		if err := cs.blockExec.ValidateBlock(cs.state, cs.ProposalBlock); err != nil {
-			panic(fmt.Sprintf("precommit step; +2/3 prevoted for an invalid block: %v", err))
+			panic(fmt.Sprintf("precommit step; +2/3 prevoted for an invalid block: %v; relocking", err))
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state.go:1652]: No. A panic unwinds the stack (it will run "defer"s tho)

File: internal/consensus/state.go [hunk 12]
```diff
@@ -2398,7 +2461,7 @@ func (cs *State) signVote(
 		ValidatorIndex:   valIdx,
 		Height:           cs.Height,
 		Round:            cs.Round,
-		Timestamp:        cs.voteTime(),
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state.go:2401]: Let's wait for all the 2022 code to land on `feature/pbts` to discuss improvements (the change you are suggesting may be coming up in a further cherry-pick)

[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state.go:2401]: Actually, I looked into the implementation of `voteTime()` and it looks like we'll have to resurrect it here when we tackle the compatibility mode (i.e., #2063)

File: internal/consensus/state.go [hunk 13]
```diff
@@ -2424,27 +2487,6 @@ func (cs *State) signVote(
 	return vote, err
 }
 
-func (cs *State) voteTime() time.Time {
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state.go:2427]: Yes. I just realized. This should be done as part of #2063

File: internal/consensus/state_test.go
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state_test.go:1]: The changes landed here as part of the cherry-pick process. Goal was reusing William's thinking at the time. All I did was choose which of his refactorings I wanted. Maybe I shouldn't have taken on the parameter change in `signAddVotes`, `validatePrevote`, `validatePrecommit`, etc. But now the work is done

[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state_test.go:1]: Thanks for reviewing this file BTW. I know it is the most cumbersome to review across the whole cherry-pick of tendermint/tendermint#7605

File: internal/consensus/state_test.go [hunk 14]
```diff
@@ -87,15 +87,16 @@ func TestStateProposerSelection0(t *testing.T) {
 	pv, err := cs1.privValidator.GetPubKey()
 	require.NoError(t, err)
 	address := pv.Address()
-	if !bytes.Equal(prop.Address, address) {
-		t.Fatalf("expected proposer to be validator %d. Got %X", 0, prop.Address)
-	}
+	require.Truef(t, bytes.Equal(prop.Address, address), "expected proposer to be validator %d. Got %X", 0, prop.Address)
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state_test.go:90]: I re-checked the diff-of-diffs. The reason is that this is how it was done in the original commit (it was inconsistent then).
As the last commits of this PR are anyway manual, I will make it consistent in another commit.

File: internal/consensus/state_test.go [hunk 15]
```diff
@@ -1313,25 +1310,19 @@ func TestStateLockMissingProposalWhenPOLForLockedBlock(t *testing.T) {
...
-	signAddVotes(cs1, types.PrevoteType, blockHash, blockParts, false, vs2, vs3, vs4)
+	signAddVotes(cs1, types.PrevoteType, cs1.state.ChainID, blockID, false, vs2, vs3, vs4)
 
 	ensurePrecommit(voteCh, height, round)
 
-	// the validator precommits the block, because it matches its locked block,
-	// maintains the same locked block and updates its locked round
-	validatePrecommit(t, cs1, round, 1, vss[0], blockHash, blockHash)
-
-	// NOTE: this behavior is inconsistent with Tendermint consensus pseudo-code.
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state_test.go:1324]: What do you mean? I see nothing removed here...

[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state_test.go:1324]: Oh, I realise, you may have been looking at the diff-of-diffs. There, this hunk would appear as "absent" in the original patch (as this UT was added by you much later and it didn't exist by the time the tendermint/tendermint#7605 landed on `main`.
If that's what captured your attention, you should notice that there are no "+" or "-" in front of these lines, and so it's just "context" that should be ignored while reviewing a cherry-pick.

[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state_test.go:1324]: Oh, wait, seems like I removed it _after_ the cherry-pick, while fixing the unit tests. Checking why....

[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state_test.go:1324]: I remember why now.
The PBTS implementation has changed the behavior: look at lines 1322 (left) and 1316 (right). We are now prevoting `nil` as your "NOTE" was saying the pseudocode does. Hence I removed it as I understand the behavior is not consistent with the pseudo code.
Could you please double-check in the production code changes that what I just said is true?

File: internal/consensus/state_test.go [hunk 16]
```diff
@@ -2821,83 +2835,6 @@ func TestResetTimeoutPrecommitUponNewHeight(t *testing.T) {
 		"triggeredTimeoutPrecommit should be false at the beginning of each height")
 }
 
-//------------------------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state_test.go:2824]: They were removed in the original commit. So I imagine there was a reason for that.
I'll try to "revive" them, but if they fail, I won't spend much time or thoughts on it.

[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state_test.go:2824]: Oh, I see they had been commented out, both on `v0.36.x` and on `main` for a _very_ long time. Chances are that they are "rotten" as they didn't run for a very very long time.

[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state_test.go:2824]: I did some git archaeology and they have been commented out since Dec 2015 😮

File: internal/state/validation.go [hunk 17]
```diff
@@ -119,13 +119,6 @@ func validateBlock(state State, block *types.Block) error {
 				state.LastBlockTime,
 			)
 		}
-		medianTime := MedianTime(block.LastCommit, state.LastValidators)
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/state/validation.go:122]: Yes, will be done as part of #2063

File: internal/store/store_test.go [hunk 18]
```diff
@@ -761,7 +762,8 @@ func TestPruneBlocks(t *testing.T) {
 	assert.EqualValues(t, 1500, bs.Height())
 	assert.EqualValues(t, 1500, bs.Size())
 
-	state.LastBlockTime = time.Date(2020, 1, 1, 1, 0, 0, 0, time.UTC)
+	state.LastBlockTime, err = time.Parse(time.RFC3339, "2035-01-16T13:00:00Z")
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/store/store_test.go:765]: This test was failing, and I realised the logic of the test needs a timestamp in the future, rather than in the past.
@jmalicevic can you confirm?

File: spec/abci/abci++_methods.md (hunk 6 above)
Comment (sergio-mena) [reply spec/abci/abci++_methods.md:757]: Thanks @glnro, that is true. Let me revert it and add it in a different PR

File: spec/core/data_structures.md
[the remark this answers is not in the document]
Comment (sergio-mena) [reply spec/core/data_structures.md:1]: Any issue number we can include here (so there's a "mention" note pointing to this comment)?

File: spec/core/data_structures.md [hunk 19]
```diff
@@ -9,38 +9,43 @@ Here we describe the data structures in the CometBFT blockchain and the rules fo
...
+  - [Execution](#execution)
+  - [Header](#header)
+  - [Version](#version)
+  - [BlockID](#blockid)
+  - [PartSetHeader](#partsetheader)
+  - [Part](#part)
+  - [Time](#time)
+  - [Data](#data)
+  - [Commit](#commit)
+  - [ExtendedCommit](#extendedcommit)
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply spec/core/data_structures.md:22]: This is passing linter, so I'll leave it as is

File: spec/core/data_structures.md [hunk 20]
```diff
@@ -485,11 +490,24 @@ func SumTruncated(bz []byte) []byte {
...
+| vote_extensions_enable_height | int64 | The ABCI application version. | 1            |
+
 ### VersionParams
 
-| Name        | Type   | Description                   | Field Number |
-|-------------|--------|-------------------------------|--------------|
-| app_version | uint64 | The ABCI application version. | 1            |
+| Name        | Type   | Description                                       | Field Number |
+| ----------- | ------ | ------------------------------------------------- | ------------ |
+| app_version | uint64 | The height where vote extensions will be enabled. | 1            |
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply spec/core/data_structures.md:503]: Indeed, thanks for pointing out (this is the obvious cherry-picking mistake we saw during the pair-programming session a couple of weeks ago)

File: types/params.go (hunk 2 above)
Comment (sergio-mena) [reply types/params.go:146]: @glnro I agree, but some work was already done as part of tendermint/tendermint#7202.
I've just added a reference to tendermint/tendermint#7202 in the task forward-porting default values for the synchrony params, so we can tackle it as part of that task.

File: spec/core/data_structures.md [hunk 21]
```diff
@@ -485,11 +490,24 @@ func SumTruncated(bz []byte) []byte {
...
+
+### SynchronyParams
+
+| Name          | Type                                                                                                                               | Description                                                                                                             | Field Number |
+| ------------- | ---------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------- | ------------ |
+| message_delay | [google.protobuf.Duration](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Duration) | Bound for how long a proposal message may take to reach all validators on a network and still be considered valid.      | 1            |
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply spec/core/data_structures.md:509]: I see. I'll fix the anchor part of the link.

File: internal/consensus/state.go [hunk 22]
```diff
@@ -1159,29 +1173,29 @@ func (cs *State) enterPropose(height int64, round int32) {
 
 	// Nothing more to do if we're not a validator
 	if cs.privValidator == nil {
-		logger.Debug("node is not a validator")
+		logger.Debug("propose step; not proposing since node is not a validator")
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state.go:1176]: We're "entering propose step". That applies for all nodes (proposer, other validators, and other full nodes).
The log line is just saying we're in that step, but we're not a validator

File: internal/consensus/replay_test.go [hunk 23]
```diff
@@ -393,7 +394,7 @@ func setupChainWithChangingValidators(t *testing.T, name string, nBlocks int) (*
 	}
 	ensureNewProposal(proposalCh, height, round)
 	rs = css[0].GetRoundState()
-	signAddVotes(css[0], types.PrecommitType, css[0].state.ChainID, types.BlockID{Hash: rs.ProposalBlock.Hash(), PartSetHeader: rs.ProposalBlockParts.Header()}, true, vss[1:nVals]...)
+	signAddVotes(css[0], types.PrecommitType, chainID, types.BlockID{Hash: rs.ProposalBlock.Hash(), PartSetHeader: rs.ProposalBlockParts.Header()}, true, vss[1:nVals]...)
```
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/replay_test.go:397]: Good point. I didn't do it as this is how it was done in the original commit. Let me see if I can improve that

**Review by glnro — APPROVED** [review]

File: internal/store/store_test.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/store/store_test.go:765]: After a discussion with @jmalicevic, we agree this change is needed because, before PBTS, the blocks' timestamps were in year 1 AD, so by 2020 any evidence had expired. With PBTS, the blocks created here are getting timestamps in the present, so 2020 is its past which made evidences to not expire.

However, together with @jmalicevic, we have improved the fix in order not to hard-code a date in the future. New commit coming up.

File: spec/core/data_structures.md
[the remark this answers is not in the document]
Comment (cason) [reply spec/core/data_structures.md:1]: #2110

File: internal/consensus/state_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (cason) [reply internal/consensus/state_test.go:1324]: You're right, as per https://github.com/tendermint/tendermint/issues/6850.

File: internal/consensus/state_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (cason) [reply internal/consensus/state_test.go:2824]: Great, lets remove unused code :)

File: internal/consensus/state.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (cason) [reply internal/consensus/state.go:1176]: I maintain my comment. If you are not a validator, you stop here. Namely, you don't propose if you are at your turn (of course not, as you are not a proposer), and you don't start the propose timeout, regardless being the proposer or not.

File: internal/consensus/state.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (cason) [reply internal/consensus/state.go:1372]: Yes, I see that, although L1378 could be changed as well.  I also realized that in line 1618 (precommit) we do the same.

The point here is that a block backed by 2/3+ prevotes is good. So we would not need to perform all checks for it. And the `timely` check is performed on the proposal.

 This is a change in the implementation of the protocol. It in general assumes that receiving a `Block` is enough for prevoting/precommiting it. This because the propagation of Proposal and Block are independent. This discussion however is more complex, should I open an issue for that?

File: .changelog/unreleased/features/7382-new-ts-validation.md
[the remark this answers is not in the document]
Comment (cason) [reply .changelog/unreleased/features/7382-new-ts-validation.md:1]: This should be reconsider as per https://github.com/cometbft/cometbft/issues/2063.

File: internal/consensus/state.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state.go:1176]: I still don't see what the problem is. Can you please suggest which changes you are expecting here?

File: internal/consensus/state.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (sergio-mena) [reply internal/consensus/state.go:1372]: Please do ahead and open an issue. Once you're done, please add a reference to that issue in the list of "Additional Changes/Revisions" in #1731

File: internal/consensus/state.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (cason) [reply internal/consensus/state.go:1176]: I would keep the original content, or use something like "skipping since node is not a validator".

File: internal/consensus/state.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (cason) [reply internal/consensus/state.go:1372]: #2119
